clang/lib/CIR/CodeGen/CIRGenCXXABI.cpp

```diff
@@ -0,0 +1,45 @@
+//===----- CirGenCXXABI.cpp - Interface to C++ ABIs -----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This provides an abstract class for C++ code generation. Concrete subclasses
+// of this implement code generation for specific C++ ABIs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "CIRGenCXXABI.h"
+#include "CIRGenFunction.h"
+
+#include "clang/AST/Decl.h"
+#include "clang/AST/GlobalDecl.h"
+
+using namespace clang;
+using namespace clang::CIRGen;
+
+CIRGenCXXABI::~CIRGenCXXABI() {}
+
+void CIRGenCXXABI::buildThisParam(CIRGenFunction &cgf,
+                                  FunctionArgList &params) {
+  const auto *md = cast<CXXMethodDecl>(cgf.curGD.getDecl());
+
+  // FIXME: I'm not entirely sure I like using a fake decl just for code
```

Comment:
Yeah, this seems pretty awful.  we should probably figure out what info we need from the decl, and just save THAT instead.

Reply:
That comment has been in the code since 2010. I'm starting to think no one is looking for a better way to do it. 😄

Reply:
Ah, shoot :D  Welp!  IMO CIR should probably model 'member function' some day, but if this is that old/preexisting, we can just live with it.

I'm generally happy with this patch, but hoping the ot hers can do a more thorough review.

Reply:
Ugh, this isn't great, and it's not the only instance in CodeGen/CIRGen where clang current creates AST nodes on the fly - one neat side effect of this is to rely on the pre-existing `ImplicitParamDecl` codegen, so it doesn't needed to be duplicated here (which usually involves handling the scalar/agg/complex versions), but in theory we could hoist the implementation of visitors into common helpers that could be used directly here.

I'm fine leaving this as-is for now.